zdb: cleanup and improve cachefile handling

ixhamza opened the pull request:
### Motivation and Context
- Automatically detect cache file; otherwise, force import.

### Description
- If a pool is created with the cache file located in a non-default path `/etc/default/zpool.cache`, removed, or the cachefile property is set to none, zdb fails to show the pool unless we specify the cache file or use the `-e` option. This PR automates this process.

### How Has This Been Tested?
CI Tests and tested with the following five scenarios. Without this patch, only scenario 1 works without specifying additional arguments:
1. Default cache file path `/etc/zfs/zpool.cache`, and the file exists.
2. Default cache file path `/etc/zfs/zpool.cache`, and the file does not exist.
3. Custom cache file path `/data/zfs/zpool.cache`, and the file exists.
4. Custom cache file path `/data/zfs/zpool.cache`, and the file does not exist.
5. Cachefile property set to `none`.

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

amotin commented on cmd/zdb/zdb.c:
```diff
@@ -9129,6 +9137,31 @@ main(int argc, char **argv)
+		zpool_handle_t *zhp;
+		char prop[ZFS_MAXPROPLEN];
+		zprop_source_t srctype = ZPROP_SRC_LOCAL;
+		if ((g_zfs = libzfs_init()) && (zhp = zpool_open_canfail(g_zfs,
+		    target)) && !zpool_get_prop(zhp, ZPOOL_PROP_CACHEFILE, prop,
+		    sizeof (prop), &srctype, B_FALSE) && prop[0] != '\0') {
```
Parts like that I prefer to wrap on logical boundaries, like `&&` here. I think it would be more readable.

Also, since we are trying to free allocated memory in first patch, shouldn't somewhere here be called zpool_close()?

amotin commented on cmd/zdb/zdb.c:
```diff
@@ -9214,9 +9246,17 @@ main(int argc, char **argv)
+		/*
+		 * If path is not provided, search in /dev
+		 */
+		if (searchdirs == NULL && nsearch == 0) {
```
Do we really need to check both or just searchdirs be enough?

ixhamza commented:
> When cachefile property is set to none, do [-d] or [-O] options require the -e option? or is that automated as well?

@akashb-22 - Considering `libzfs` is able to read the `cachefile` property correctly, it should be automated.

akashb-22 commented:
I see some disparity when I use the pool-name and dataset name as target.
```
# zpool get cachefile
NAME       PROPERTY   VALUE      SOURCE
pool-oss5  cachefile  none       local

# zdb -d pool-oss5
Dataset mos [META], ID 0, cr_txg 4, 5.62M, 171 objects
Dataset pool-oss5/ost5 [ZPL], ID 387, cr_txg 67, 1.20G, 7 objects
Dataset pool-oss5 [ZPL], ID 54, cr_txg 1, 1024K, 6 objects
Verified large_blocks feature refcount of 1 is correct
Verified large_dnode feature refcount of 1 is correct
Verified sha512 feature refcount of 0 is correct
Verified skein feature refcount of 0 is correct
Verified edonr feature refcount of 0 is correct
Verified userobj_accounting feature refcount of 1 is correct
Verified encryption feature refcount of 0 is correct
Verified project_quota feature refcount of 1 is correct
Verified redaction_bookmarks feature refcount of 0 is correct
Verified redacted_datasets feature refcount of 0 is correct
Verified bookmark_written feature refcount of 0 is correct
Verified livelist feature refcount of 0 is correct
Verified zstd_compress feature refcount of 0 is correct
Verified zilsaxattr feature refcount of 1 is correct
Verified blake3 feature refcount of 0 is correct
Verified redaction_list_spill feature refcount of 0 is correct
Verified device_removal feature refcount of 0 is correct
Verified indirect_refcount feature refcount of 0 is correct

# zdb -d pool-oss5/ost5
failed to hold dataset 'pool-oss5/ost5': No such file or directory
zdb: can't open 'pool-oss5/ost5': No such file or directory

# zdb -d pool-oss5/ost5 -e
Dataset pool-oss5/ost5 [ZPL], ID 387, cr_txg 67, 1.20G, 7 objects
```
I think the dataset as the target doesn't go well here maybe, I'm not sure, the same with -e seems to be working fine. 
@ixhamza  is this expected?

ixhamza commented:
@akashb-22 - I have made some adjustments to automate the scenario you mentioned. Can you give it a try now?

akashb-22 reviewed (approved):
LGTM.

behlendorf commented on cmd/zdb/zdb.c:
```diff
@@ -9112,23 +9137,63 @@ main(int argc, char **argv)
+		free(pname);
+	}
+
+	kernel_init(SPA_MODE_READ);
```
Moving the `kernel_init()` call here seems like the most likely cause the ASSERT we're now hitting reliably in `alloc_class/alloc_class_012_pos`.

ixhamza replied:
@behlendorf - Today I got a chance to dig deeper, and it appears that it only fails when ZFS is compiled from RPM packaging, consistently triggering this ASSERT: https://github.com/openzfs/zfs/blob/master/module/zcommon/zfs_fletcher.c#L429. Interestingly, even if I revert the entire patch and just keep the inclusion of `libzfs.la` in the `cmd/zdb/Makefile`, the issue still reproduces. I don't have much experience with RPM packaging, but I'll try to figure it out. Any suggestions would be appreciated.

behlendorf commented on cmd/zdb/zdb.c:
```diff
@@ -9458,17 +9532,22 @@ main(int argc, char **argv)
+		libzfs_fini(g_zfs);
+
+	if (kernel_init_done)
+		kernel_fini();
```
> When encountering an error or early return, kernel_fini is not called.

Can you split the `kernel_fini()` in to its own patch and open a new PR for it.  There's no reason this needs to be blocked by the larger cache file detection improvements.

ixhamza replied:
Thanks. Opened https://github.com/openzfs/zfs/pull/16152.

ixhamza commented:
@behlendorf - I have updated the patch according to your suggestions and rebased it with master branch. The `alloc_class` tests, which were previously failing, are now passing. However, the remaining failing test cases seem to be unrelated.

behlendorf reviewed (approved):
Thanks for restructuring this.  Those CI failures look unrelated to me but I'll resubmit those failed runs to make sure.
